File: Programmers/Level2/다리를지나는트럭.java
package week01;

import java.util.LinkedList;
import java.util.Queue;

public class 다리를지나는트럭 {

	public static void main(String[] args) {
		Solution s = new Solution();
		int bl = 100;
		int we = 100;
		int[] tr_we = {10};
		System.out.println(s.solution3(bl,we,tr_we));

	}

}

class Solution {
    public int solution3(int bridge_length, int weight, int[] truck_weights) {
    		Queue<Integer> bridge = new LinkedList<>();
    		Queue<Integer> readyTruck = new LinkedList<>();
    		
            //초기 다리 
            for(int i = 0 ; i < bridge_length ; i++)
            {
                bridge.offer(0);
            }
           
            //대기 트럭
            for(int i = 0; i < truck_weights.length; i++) {
            	readyTruck.offer(truck_weights[i]);
            }
            
            int time = 0;
            int totalWeight = 0;
            while(!bridge.isEmpty()) {
            	totalWeight -= bridge.peek();
            	bridge.poll();
            	if(!readyTruck.isEmpty()) {
            		if((totalWeight + readyTruck.peek()) <= weight) {
            			bridge.offer(readyTruck.peek());
            			totalWeight += readyTruck.peek();
                		readyTruck.poll();
                		
            		}
            		else
            			bridge.offer(0);
            		
            	}
            	else {
            		time += bridge_length;
            		break;
            
            	}
            	time++;
            }
            return time;
    }
}

File: Programmers/다리를지나는트럭.js
function solution(bridge_length, weight, truck_weights) {
  var answer = 0;
  const queue = Array(bridge_length).fill(0);

  while (queue.length) {
    console.log(queue);
    queue.shift();
    console.log(queue);
    if (truck_weights.length) {
      if (queue.reduce((a, b) => a + b, 0) + truck_weights[0] <= weight) {
        queue.push(truck_weights.shift());
      } else {
        queue.push(0);
      }
    }
    answer++;
  }

  return answer;
}

console.log(solution(2, 10, [7, 4, 5, 6])); // 8
// console.log(solution(100, 100, [10])); // 101
// console.log(solution(100, 100, [10, 10, 10, 10, 10, 10, 10, 10, 10, 10])); // 110
